src/tutorial/hello-world/README.md

```diff
@@ -1,2 +1,245 @@
+
+#define macro MAIN() = takes(0) returns(0) {
+    
+    // Get the function selector
+    0x00 
+    calldataload 
+    0xE0 
+    shr
+    
+    // Jump to the implementation of the ADD_TWO function if the calldata matches the function selector
+    __FUNC(addTwo) addTwo eq jumpi  
+
+    addTwo:
+        ADD_TWO()
+}
+
+#define macro ADD_TWO() = takes(0) returns(0) {
+    0x04 calldataload     // load first 32 bytes onto the stack - number 1
+    0x24 calldataload     // load second 32 bytes onto the stack - number 2
+    add                   // add number 1 and 2 and put the result onto the stack
+    
+    0x00 mstore           // place the result in memory
+    0x20 0x00 return      // return the result
+}
+```
+
+The first modifications we make will be within the ADD_TWO macro. On lines 1 and 2 we will shift the calldata offset by 4 bytes for both numbers, this is due to the 4 byte function selector that will be prepended to the calldata value. 
+
+Our `MAIN` macro has changed drastically. 
+The first 4 lines are concerned with isolating the function selector from the calldata.
+1. `0x00` pushed `[0]` onto the stack
+2. `calldataload` takes `[0]` as input and pushes the first 32 bytes of calldata onto the stack
+3. `0xE0` pushes `[224]` onto the stack. This magic number represents 256 bits - 32 bits (32 bytes - 4 bytes). 
+4. When followed by the shr this will place the function selector onto the stack
+
+The following lines will match the function selector on the stack and then jump to the code location where that code exists. Huff handles generating all jump logic for you.
+
+Under the hood the ADD_TWO() macro bytecode will be inlined of ADD_TWO() in the main macro. 
+
+Now you should be able to use libraries like ethers, or other contracts to call your huff contract! 
+
+We hope this now gives you a good understanding of the main concepts and all of the boiler plate you need to get started in huff!
+
+**Hello, world!**
+Earlier we mentioned how `"Hello, world!"` is quite an advanced concept for huff. The reason being that we have to clear understanding of EVM string encoding. As strings are dynamic types it is not as simple as returning the UTF-8 values for `"Hello, world!"` (`0x48656c6c6f2c20776f726c6421`). In the abi standard, dynamic types are encoded in 3 parts.
+1. The offset of the dynamic data.
+2. The length of the dynamic data.
+3. The values of the dynamic data.
+
+Each part will look as follows when returning the string `"Hello, world!"`:
+```
+0x0000000000000000000000000000000000000000000000000000000000000020 // The location of the "Hello, world!" data (dynamic type).
+```
+```
+0x000000000000000000000000000000000000000000000000000000000000000d // The length of "Hello, world!" in bytes
+```
+```
+0x48656c6c6f2c20776f726c642100000000000000000000000000000000000000 // Value "Hello, world!"
+```
+
+Encoding dynamic values takes alot of work!! In order to return `"Hello, world!"` we must return 96 bytes!.
```

Comment:
i understand what u did below in the code, and this is the canonical way to do it, but idt its explained that well here to a beginner.  specifically, when we concat the `0d` to the beginning of `48..21` and then use offset 0x2d.  This has the effect of putting the 0d in the lowest bits on the second word, and the rest in the high bits in the third word.

i think some more explanation around that is warranted here

Reply:
Good point, I'll split them up to make it more clear!